numpy/core/src/umath/matmul.c.src

```diff
@@ -26,17 +26,36 @@
  *****************************************************************************
  */
 
+/*
+ * -1 to be conservative, in case blas internally uses a for loop with an
+ * inclusive upper bound
+ */
+#define BLAS_MAXSIZE (NPY_MAX_INT - 1)
+
 /*
  * Determine if a non-contiguous axis has strides and length supported by blas.
  * 1. Strides must be multiples of the itemsize, and must not cause segments of
  *    the contiguous axis to overlap.
  * 2. BLAS requires sizes to fit within `int`s.
  */
-
-#define BLAS_ABLE(n, stride, itemsize) ((stride % itemsize == 0) && \
-                                  (n < NPY_MAX_INT) && (stride == itemsize || \
-                                  ((stride / itemsize >= n) && \
-                                  (stride / itemsize < NPY_MAX_INT))))
+static NPY_INLINE npy_bool
+is_blasable(npy_intp n, npy_intp byte_stride, npy_intp itemsize)
+{
+    npy_intp unit_stride = byte_stride / itemsize;
+    if (byte_stride == 0) {
+        return NPY_FALSE;
+    }
+    if (byte_stride == itemsize) {
+        return NPY_TRUE;
+    }
```

Comment:
I don't think these three lines are correct, which is why I was pushing for a distinction between "is supported by blas" and "is supported by blas if I transpose it.

Reply:
ok, will extend the function